src/commissioner/States.h

```diff
@@ -0,0 +1,840 @@
+    }
+
+protected:
+    CHIP_ERROR GetNextCandidate()
+    {
+        if (mIdx >= mRecord.get()->mNumIPs)
+        {
+            return CHIP_ERROR_NOT_FOUND;
+        }
+        Inet::InterfaceId interfaceId =
+            mRecord.get()->mAddress[mIdx].IsIPv6LinkLocal() ? mRecord.get()->mInterfaceId : Inet::InterfaceId::Null();
+#if 0
+        // TODO(#13326): connecting via TCP does not work
+        if (mRecord.get()->supportsTcp)
+        {
+            this->mCommissionee.mOperationalAddress.SetValue(Transport::PeerAddress::TCP(mRecord.get()->mAddress[mIdx], mRecord.get()->mPort, interfaceId));
+            this->mCommissionee.mMrpConfig.ClearValue();
+        }
+        else
+#endif
+        {
+            this->mCommissionee.mOperationalAddress.SetValue(
+                Transport::PeerAddress::UDP(mRecord.get()->mAddress[mIdx], mRecord.get()->mPort, interfaceId));
+            this->mCommissionee.mMrpConfig.SetValue(mRecord.get()->GetMRPConfig());
+        }
+#if CHIP_PROGRESS_LOGGING
+        char addressStr[Transport::PeerAddress::kMaxToStringSize];
+        this->mCommissionee.mOperationalAddress.Value().ToString(addressStr);
+        ChipLogProgress(Controller, "Operational identity discovered at %s", addressStr);
+#endif // CHIP_PROGRESS_LOGGING
+        ++mIdx;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR TryCase()
+    {
+        CHIP_ERROR err                        = CHIP_NO_ERROR;
+        uint16_t sessionId                    = 0;
+        Messaging::ExchangeContext * exchange = nullptr;
+        Optional<SessionHandle> session;
+        FabricInfo * fabric;
+
+        fabric = this->mCommissionee.mSystemState->Fabrics()->FindFabricWithCompressedId(
+            this->mCommissionee.mOperationalId.Value().GetCompressedFabricId());
+        VerifyOrExit(fabric != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+        mPairing = Platform::MakeShared<CASESession>();
+        VerifyOrExit(mPairing.get() != nullptr, err = CHIP_ERROR_NO_MEMORY);
+        session = this->mCommissionee.mSystemState->SessionMgr()->CreateUnauthenticatedSession(
+            this->mCommissionee.mOperationalAddress.Value(), this->mCommissionee.mMrpConfig.ValueOr(gDefaultMRPConfig));
+        VerifyOrExit(session.HasValue(), err = CHIP_ERROR_NO_MEMORY);
+        exchange = this->mCommissionee.mSystemState->ExchangeMgr()->NewContext(session.Value(), this->mPairing.get());
+        VerifyOrExit(exchange != nullptr, err = CHIP_ERROR_INTERNAL);
+        SuccessOrExit(err = this->mCommissionee.mSystemState->GetSessionIDAllocator()->Allocate(sessionId));
+        SuccessOrExit(err = mPairing.get()->EstablishSession(this->mCommissionee.mOperationalAddress.Value(), fabric,
+                                                             this->mCommissionee.mOperationalId.Value().GetNodeId(), sessionId,
+                                                             exchange, this, this->mCommissionee.mMrpConfig));
+    exit:
```

Comment:
Similar comment else where, do we need to release `exchange`?

Reply:
Correct.  Will fix.

Reply:
Donbe.

Reply:
This turned out to be more complex than first thought.  See https://github.com/project-chip/connectedhomeip/pull/13294#discussion_r780844434.

But this is now resolved in the PR.